tinygrad/runtime/ops_webgpu.py

```diff
@@ -2,62 +2,280 @@
+
+def copy_buffer_to_buffer(device, src, src_offset, dst, dst_offset, size):
+  encoder = webgpu.wgpuDeviceCreateCommandEncoder(device, webgpu.WGPUCommandEncoderDescriptor())
+  webgpu.wgpuCommandEncoderCopyBufferToBuffer(encoder, src, src_offset, dst, dst_offset, size)
+  cb = webgpu.wgpuCommandEncoderFinish(encoder, webgpu.WGPUCommandBufferDescriptor())
+  submit(device, [cb])
+  webgpu.wgpuCommandBufferRelease(cb)
+  webgpu.wgpuCommandEncoderRelease(encoder)
+
+def read_buffer(dev, buf):
+  size = webgpu.wgpuBufferGetSize(buf)
+  tmp_usage = webgpu.WGPUBufferUsage_CopyDst | webgpu.WGPUBufferUsage_MapRead
+  tmp_buffer = webgpu.wgpuDeviceCreateBuffer(dev, webgpu.WGPUBufferDescriptor(size=size, usage=tmp_usage, mappedAtCreation=False))
+  copy_buffer_to_buffer(dev, buf, 0, tmp_buffer, 0, size)
+  map_buffer(tmp_buffer, size)
+  ptr = webgpu.wgpuBufferGetConstMappedRange(tmp_buffer, 0, size)
+  void_ptr = ctypes.cast(ptr, ctypes.c_void_p)
+  byte_array = (ctypes.c_uint8 * size).from_address(void_ptr.value)
+  result = bytearray(byte_array)
+  webgpu.wgpuBufferUnmap(tmp_buffer)
+  webgpu.wgpuBufferDestroy(tmp_buffer)
+  return memoryview(result).cast("B")
+
+def submit(device, command_buffers):
+  cb_buffers_array_type = webgpu.WGPUCommandBuffer * len(command_buffers)
+  cb_buffers_array = cb_buffers_array_type(*command_buffers)
+  webgpu.wgpuQueueSubmit(webgpu.wgpuDeviceGetQueue(device), len(command_buffers), cb_buffers_array)
+
+def pop_error(device):
+  result: List[Any] = []
+
+  def cb(status, err_type, msg, i2): result[:] = [from_wgpu_str(msg)]
+
+  cb_info = create_cb_info(webgpu.WGPUPopErrorScopeCallbackInfo, webgpu.WGPUPopErrorScopeCallback, cb)
+  wgpu_wait(webgpu.wgpuDevicePopErrorScopeF(device, cb_info))
+  return result[0] if len(result) > 0 else ""
+
+def create_uniform(wgpu_device, val):
+  buf = webgpu.wgpuDeviceCreateBuffer(wgpu_device,
+    webgpu.WGPUBufferDescriptor(size=4, usage=webgpu.WGPUBufferUsage_Uniform | webgpu.WGPUBufferUsage_CopyDst))
+  write_buffer(wgpu_device, buf, 0, val.to_bytes(4, "little") if isinstance(val, int) else struct.pack('<f', val))
   return buf
 
 class WebGPUProgram:
   def __init__(self, dev, name:str, lib:bytes):
     (self.dev, self.timestamp_supported) = dev
-    self.name, self.lib, self.prg = name, lib, self.dev.create_shader_module(code=lib.decode())   # NOTE: this is the compiler
+
+    # Creating shader module
+    shader = webgpu.WGPUShaderModuleWGSLDescriptor()
+    shader.code = to_wgpu_str(lib.decode())
+    shader.chain.next = None
+    shader.chain.sType = webgpu.WGPUSType_ShaderSourceWGSL
+    module = webgpu.WGPUShaderModuleDescriptor()
+    module.nextInChain = ctypes.cast(ctypes.pointer(shader), ctypes.POINTER(webgpu.struct_WGPUChainedStruct))
+
+    # Check compiler error
+    webgpu.wgpuDevicePushErrorScope(self.dev, webgpu.WGPUErrorFilter_Validation)
+    shader_module = webgpu.wgpuDeviceCreateShaderModule(self.dev, module)
```

Comment:
Not required for this PR, but is there a way for the Compiler to be separate from the Program?